core/jobs/batch_jobs/question_migration_jobs.py

```diff
@@ -0,0 +1,256 @@
+
+        unused_put_results = (
+            question_summary_models_to_put
+            | 'Put models into datastore' >> ndb_io.PutModels()
+        )
+
+        return transformed_question_summary_job_run_results
+
+
+class AuditPopulateQuestionSummaryVersionOneOffJob(base_jobs.JobBase):
+    """Job that audits PopulateQuestionSummaryVersionOneOffJob."""
+
+    @staticmethod
+    def _transform_question_summary(
+        question_summary_id: str,
+        question_summary_model: question_models.QuestionSummaryModel
+    ) -> result.Result[Tuple[str, question_domain.QuestionSummary],
+        Tuple[str, Exception]
+    ]:
+        """Transform question summary model into question summary object.
+
+        Args:
+            question_summary_id: str. The id of the question summary.
+            question_summary_model: QuestionSummaryModel. The question model
+                to migrate.
+
+        Returns:
+            Result((str, QuestionSummary), (str, Exception)). Result containing
+            tuple that consist of question ID and either question summary
+            object or Exception. Question summary object is returned when the
+            migration was successful and Exception is returned otherwise.
+        """
+        try:
+            question_summary = (
+                question_services.get_question_summary_from_model(
+                question_summary_model
+                )
+            )
+            question_summary.validate()
+        except Exception as e:
+            logging.exception(e)
+            return result.Err((question_summary_id, e))
+
+        return result.Ok((question_summary_id, question_summary))
+
+    def run(self) -> beam.PCollection[job_run_result.JobRunResult]:
+        """Returns a PCollection of results from the question migration.
+
+        Returns:
+            PCollection. A PCollection of results from the question
+            migration.
+        """
+        question_summary_models = (
+            self.pipeline
+            | 'Get all non-deleted question summary models' >> (
+                ndb_io.GetModels(
+                    question_models.QuestionSummaryModel.get_all())
+                )
+            # Pylint disable is needed because pylint is not able to correctly
+            # detect that the value is passed through the pipe.
+            | 'Add question summary keys' >> beam.WithKeys( # pylint: disable=no-value-for-parameter
+                lambda question_summary_model: question_summary_model.id)
+        )
+
+        all_transformed_question_summary_results = (
+            question_summary_models
+            | 'Transform model' >> beam.MapTuple(
+                self._transform_question_summary)
+        )
+
+        transformed_question_summary_job_run_results = (
+            all_transformed_question_summary_results
+            | 'Generates results' >> (
+                job_result_transforms.ResultsToJobRunResults(
+                    'QUESTION SUMMARY PROCESSED'))
```

Comment:
Can we have all the steps from the above job except the "'Put models into datastore'"? (This will help us ensure all the steps in the above job are working correctly on all the models.)

@vojtechjelinek Do we have any method to write better audit jobs?

(I had tried this in past: https://github.com/oppia/oppia/blob/58c610d01bd72e313634a38d90c037ade64b3ad2/core/jobs/batch_jobs/suggestion_migration_jobs.py#L203 you can use it if we don't have any general way to write audit jobs.)

Reply:
@DubeySandeep Generally, we use the method here to write audit jobs. The common code is moved to a transform which is then used in migration and audit jobs.

https://github.com/oppia/oppia/blob/58c610d01bd72e313634a38d90c037ade64b3ad2/core/jobs/batch_jobs/topic_migration_jobs.py#L355

Reply:
But I don't see the "actual job" and the "audit job" has all the common steps. I think the audit job should contain all the steps of the "Actual job" except the "put model" step. Can we do something to avoid duplicate code and ensure all the steps are performed in the audit job?

Reply:
Yes I will add the steps from the "actual job" to the audit job. I haven't used the method (the one I described above) here since this a one off job and would be removed later. That comment was a general answer to the question you asked before.

> @vojtechjelinek Do we have any method to write better audit jobs?

Reply:
Ah, I see. Thanks for explaining it @JayVivarekar

Reply:
> Can we do something to avoid duplicate code and ensure all the steps are performed in the audit job?

is this resolved?

Reply:
Yes